File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
info = dict()
def solution(orders, course):
    answer = []
    orders_list = []
    for order in orders:
        arr = list(order)
        arr.sort()
        orders_list.append(arr)

    
    for cnt in course:
        # info = dict()
        max_cnt = 0
        for order in orders_list:
            dfs(0, "", order, cnt)
        
        if info:
            max_value = max(info.values())
            if max_value > 1:
                for k, v in info.items():
                    if v == max_value:
                        answer.append(k)
        info.clear()      
    answer.sort()
    return answer

def dfs(depth, menu, chars, total):
    # print(depth, menu, chars, total)
    if total == len(menu):
        if menu in info:
            info[menu] += 1
        else:
            info[menu] = 1
        return
    
    if depth >= len(chars):
        return
    
    dfs(depth + 1, menu + chars[depth], chars, total)
    dfs(depth + 1, menu, chars, total)

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.Map;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

class Solution {
    static ArrayList<String> answerList = new ArrayList<>();
    static Map<String, Integer> map = new HashMap<>();
    
    public String[] solution(String[] orders, int[] course) {
        String[] answer = {};
        
        String[] newOrders = new String[orders.length];
        for (int i = 0; i < orders.length; i++) {
            char[] tmp = orders[i].toCharArray();
            Arrays.sort(tmp);
            newOrders[i] = new String(tmp);
        }
        
        for (int cnt : course) {
            for (String order : newOrders) {
                dfs(0, "", order, cnt);
            }
            
            if (!map.isEmpty()) {
                ArrayList<Integer> countList = new ArrayList<>(map.values());
                int max = Collections.max(countList);
                if (max > 1) {
                    for (String key : map.keySet()) {
                        if (map.get(key) == max) {
                            answerList.add(key);
                        }
                    }
                }
            }
            map.clear();
        }
        Collections.sort(answerList);
        answer = new String[answerList.size()];
        for (int i = 0; i < answerList.size(); i++) {
            answer[i] = answerList.get(i);
        }
        
        return answer;
    }
    
    static void dfs(int depth, String course, String chars, int total) {
        if (course.length() == total) {
            map.put(course, map.getOrDefault(course, 0) + 1);
            return;
        }
        
        if (depth >= chars.length()) {
            return;
        }
        
        dfs(depth+1, course, chars, total);
        dfs(depth+1, course + chars.charAt(depth), chars, total);
    }
}